### DISCUSSION TITLE
What do NaN uncertainties mean?

### DISCUSSION THREAD

**Discussion by Melissa-Iacovidou** [JuliaPhysics/Measurements.jl#91] [Q&A]
I'm using the Measurements.jl package to error propagate through multiple calculations including integrations (quadgk, cuba). It works well but there is an instance where I get results with ± NaN as uncertainties. Why is this happening?

**Comment by giordano** [comment]
Probably you get a `NaN` somewhere in your calculations (`0/0` or `Inf - Inf`) which poisons the uncertainty.  What's the physical meaning of a `NaN` uncertainty?  Probably not much.

Would you be able to share your code?

**Reply by Melissa-Iacovidou** [reply]
For example:

```julia
julia> using Measurements, QuadGK

julia> a = 0.005 ± 0.0005
0.005 ± 0.0005

julia> b = 0.1 ± 0.005
0.1 ± 0.005

julia> quadgk(t -> exp(-(a * exp(b * t))), 0, +Inf)
(47.2609545858445 ± NaN, 6.257142884154728e-7 ± NaN)
```

**Reply by giordano** [reply]
Basically that comes from the fact the function is evaluated at infinite:
```julia
julia> a = 0.005 ± 0.0005
0.005 ± 0.0005

julia> b = 0.1 ± 0.005
0.1 ± 0.005

julia> exp(-(a * exp(b * Inf)))
0.0 ± NaN
```
Unfortunately, I don't think there is much we can do about this, given how the linear error propagation theory works

**Reply by giordano** [reply]
More specifically
```julia
julia> -(a * exp(b * Inf))
-Inf ± Inf
```
and then taking the exponential of this quantity isn't good for the uncertainty (well, it was already infinite):
```julia
julia> exp(-(a * exp(b * Inf)))
0.0 ± NaN
```

**Reply by Melissa-Iacovidou** [reply]
Okay, so what I shared isn't exactly what I'm having trouble with, it is a simplification of what I am doing. I checked what you said with my actual code and despite having infinity there, in some cases I don't get NaN and in others I do. 

This is part of the code:
`
for i in collect(1:65)
       println(quadgk(t -> σ * exp(-σ*t) * exp(-((a * exp(i * b) * (-1 + exp(b * t))) /b)), 0, +Inf)[1])
end
`

So why is it for some i's the uncertainty is NaN and for others it gives a value?

**Reply by giordano** [reply]
What's `σ`?

**Reply by Melissa-Iacovidou** [reply]
oops sorry forgot to mention, it's not really important but let's say 0.1

**Reply by giordano** [reply]
I think the integrand function may not be always evaluated exactly on `t = Inf`

**Reply by Melissa-Iacovidou** [reply]
ok thanks, i'll look into the quadgk package!

**Reply by giordano** [reply]
Frankly, I think there is little to do, at least within `Measurements.jl`: when you get pathological operations like `exp(Inf ± Inf)` this is numerically what happens.  In principle you may have more reasonable results with MonteCarlo propagation (like [`MonteCarloMeasurements.jl`](https://github.com/baggepinnen/MonteCarloMeasurements.jl)), but that's a completely different way to propagate uncertainties (and not sure whether it solves the issue at infinite values)

**Reply by Melissa-Iacovidou** [reply]
Actually `exp(Inf ± Inf)` gives Inf ± Inf, it's `exp(- Inf ± Inf)` that gives 0.0 ± NaN. Given the error propagation theory, why isn't that just 0.0 ± 0.0?

**Reply by giordano** [reply]
The error of `exp(x_0)` is given by the product of the derivative of `exp(x)`, evaluated at `x = x_0`, times the error of `x_0`: `exp(-Inf) = 0` and the error is `Inf`: `0 * Inf = NaN`

**Reply by Melissa-Iacovidou** [reply]
You are right, thanks for all the help!

**Reply by giordano** [reply]
I'm sorry for not having being able to solve your issue, but unfortunately things get numerically tricky at pathological values :disappointed:  even though in some cases limits may be well defined, mathematically, computers are dumb and `0 * Inf` always results in `NaN` (unless one adds a more complex algebra to the system...)

**Reply by Melissa-Iacovidou** [reply]
No worries, it seems like if I numerically integrate until say 10^3 instead of Inf, the results are essentially the same, and I get an actual error. I checked the value of my integral from 10^3 to Inf and it gives me zero, so the contribution seems to be negligible. I think this could be a temporary fix at the moment. Might consider doing a log transformation or something.

**Reply by giordano** [reply]
Yeah, if the integral between your threshold and `Inf` is less than the uncertainty you'd get, it does make sense to stop earlier!
